amaranth/hdl/_time.py

```diff
@@ -0,0 +1,249 @@
+
+    @property
+    def nanoseconds(self):
+        return self._femtoseconds / 1_000_000
+
+    @property
+    def picoseconds(self):
+        return self._femtoseconds / 1_000
+
+    @property
+    def femtoseconds(self):
+        return self._femtoseconds
+
+    def _check_reciprocal(self):
+        if self._femtoseconds == 0:
+            raise ZeroDivisionError("Can't calculate the frequency of a zero period")
+        elif self._femtoseconds < 0:
+            raise ValueError("Can't calculate the frequency of a negative period")
+
+    @property
+    def hertz(self):
+        self._check_reciprocal()
+        return 1_000_000_000_000_000 / self._femtoseconds
+
+    @property
+    def kilohertz(self):
+        self._check_reciprocal()
+        return 1_000_000_000_000 / self._femtoseconds
+
+    @property
+    def megahertz(self):
+        self._check_reciprocal()
+        return 1_000_000_000 / self._femtoseconds
+
+    @property
+    def gigahertz(self):
+        self._check_reciprocal()
+        return 1_000_000 / self._femtoseconds
+
+    def __lt__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds < other._femtoseconds
+
+    def __le__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds <= other._femtoseconds
+
+    def __eq__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds == other._femtoseconds
+
+    def __ne__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds != other._femtoseconds
+
+    def __gt__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds > other._femtoseconds
+
+    def __ge__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds >= other._femtoseconds
+
+    def __hash__(self):
+        return hash(self._femtoseconds)
+
+    def __bool__(self):
+        return bool(self._femtoseconds)
+
+    def __neg__(self):
+        return Period(fs = -self._femtoseconds)
+
+    def __pos__(self):
+        return self
+
+    def __abs__(self):
+        return Period(fs = abs(self._femtoseconds))
+
+    def __add__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return Period(fs = self._femtoseconds + other._femtoseconds)
+
+    def __sub__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return Period(fs = self._femtoseconds - other._femtoseconds)
+
+    def __mul__(self, other):
+        if not isinstance(other, numbers.Real):
+            return NotImplemented
+        return Period(fs = self._femtoseconds * other)
+
+    def __rmul__(self, other):
```

Comment:
this is subtly, but badly broken.

delegating via calling `*` will go into an infinite recursion when there are two different classes involved that have similar `NotImplemented` handling, because it will repeat the `__mul__`/`__rmul__` resolution dance. for an example, run this python program:

```py
class A:
    def __mul__(self, other):
        if not isinstance(other, A):
            return NotImplemented
        return 1

    def __rmul__(self, other):
        return self * other

class B:
    def __mul__(self, other):
        if not isinstance(other, B):
            return NotImplemented
        return 1

    def __rmul__(self, other):
        return self * other

print(A() * B())
```

what you should do instead is simply:

```
__rmul__ = __mul__
```